### ISSUE TITLE
impossible to differentiate over integer array

### ISSUE THREAD

**Issue by lliill** [JuliaDiff/DiffResults.jl#10]
`l(xu) = sum(sqrt(i^2 + 1) for i in xu)`
`xu = Vector(1:6)`
`result = DiffResults.HessianResult(xu)`

`result = ForwardDiff.hessian!(result, l, xu)` will give
`InexactError: Int64(Int64, 22.11744686954993)`

Problem will be solve if `xu = Vector{Float64}(1:6)`

**Comment by gdalle** [comment]
You're right, this gives an error, but there is a good reason. Enforcing that the result keeps the same number types is a good thing for type-stability, and that's why I don't think it would be a good idea to change that

**Comment by gdalle** [comment]
Reopening because a warning in the docs might make sense
